### PR TITLE
Force usage of NonBlocking dictionary

### PR DESCRIPTION
**Pull request by vladimirnesic1980**
Force usage of NonBlocking dictionary
# Description

# Related Issue\Feature

# How Has This Been Tested

- [ ] All unit tests pass.
- [ ] All integration tests pass.
- [ ] Manual Testing:

# Types of changes

- [ ] Docs change
- [ ] Refactoring
- [ ] Dependency upgrade
- [ ] Additional Unit Tests\Integration Tests
- [ ] Bug fix (non-breaking change which fixes an issue)
- [x] New feature (non-breaking change which adds functionality)
- [ ] Breaking change (fix or feature that would cause existing functionality to change)
- [ ] Removed no-longer used code

## Deployment Configuration Changes

- [ ] Requires deployment configuration changes as specified below and in CHANGELOG.md

# Checklist

- [ ] There are no Resharper\static code analysis errors anywhere in the solution.
- [ ] I have ONLY run a code clean-up on any files I have modified to make sure they are in the correct format and no others.
- [x] I have added tests to cover my changes.
- [ ] I have run the code and quickly verified it all works to my satisfaction.
- [ ] All new/modified code has sufficient logging to be able to diagnose what is wrong.
- [x] All new and existing tests passed.
- [ ] All new/modified public interfaces/classes have are documented with xmldoc comments.
- [x] Unreleased section of CHANGELOG.md has been updated with details of this PR.

### CODE REVIEW

File: src/FunFair.CodeAnalysis.Tests/Helpers/WellKnownMetadataReferences.cs [hunk 1]
```diff
@@ -39,5 +40,9 @@ internal static class WellKnownMetadataReferences
         public static readonly MetadataReference ConnectionInfo = MetadataReference.CreateFromFile(typeof(ConnectionInfo).Assembly.Location);
 
         public static readonly MetadataReference SuppressMessage = MetadataReference.CreateFromFile(typeof(SuppressMessageAttribute).Assembly.Location);
+
+        public static readonly MetadataReference ConcurrentDictionary  = MetadataReference.CreateFromFile(typeof(ConcurrentDictionary<int,int>).Assembly.Location);
+
+        public static readonly MetadataReference NonBlockingConcurrentDictionary  = MetadataReference.CreateFromFile(typeof(NonBlocking.ConcurrentDictionary<int,int>).Assembly.Location);
```
Comment (credfeto) [inline src/FunFair.CodeAnalysis.Tests/Helpers/WellKnownMetadataReferences.cs:46]: not sure if you can but :

```NonBlocking.ConcurrentDictionary<int,int>``` could be ```NonBlocking.ConcurrentDictionary<,>```?

File: src/FunFair.CodeAnalysis.Tests/ProhibitedClassesDiagnosticsAnalyzerTests.cs [hunk 2]
```diff
@@ -0,0 +1,159 @@
...
+{
+    public sealed class ProhibitedClassesDiagnosticsAnalyzerTests : CodeFixVerifier
+    {
+        protected override DiagnosticAnalyzer GetCSharpDiagnosticAnalyzer()
+        {
+            return new ProhibitedClassesDiagnosticsAnalyzer();
+        }
+
+        [Fact]
+        public Task AssertFalseWithMessageIsAllowedAsync()
```
Comment (credfeto) [inline src/FunFair.CodeAnalysis.Tests/ProhibitedClassesDiagnosticsAnalyzerTests.cs:18]: probaby want to give this a better name?

File: src/FunFair.CodeAnalysis/ProhibitedClassesDiagnosticsAnalyzer.cs [hunk 3]
```diff
@@ -0,0 +1,158 @@
...
+        private static INamedTypeSymbol? GetSymbol(ITypeSymbol symbol, Dictionary<string, INamedTypeSymbol> cachedSymbols)
+        {
+            string typeName = symbol.ToFullyQualifiedName();
+
+            if (cachedSymbols.ContainsKey(typeName))
+            {
+                return cachedSymbols[typeName];
+            }
+
+            return null;
```
Comment (credfeto) [inline src/FunFair.CodeAnalysis/ProhibitedClassesDiagnosticsAnalyzer.cs:142]: ```return cachedSymbols.TryGetValue(typename, out INamedTypeSymbol? symbol) ? symbol : null;```

File: src/FunFair.CodeAnalysis.Tests/ProhibitedClassesDiagnosticsAnalyzerTests.cs [hunk 4]
```diff
@@ -0,0 +1,159 @@
...
+        public Task AssertTrueForParameterInMethodDeclarationWithMessageIsBannedAsync()
+        {
+            const string test = @"
+     using System.Collections.Concurrent;
+
+     namespace ConsoleApplication1
+     {
+         class TypeName
+         {
+             void Test(ConcurrentDictionary<int,int> dictionary)
```
Comment (credfeto) [inline src/FunFair.CodeAnalysis.Tests/ProhibitedClassesDiagnosticsAnalyzerTests.cs:74]: Does this cope with constructor parameters too?

Comment (vladimirnesic1980) [reply src/FunFair.CodeAnalysis.Tests/ProhibitedClassesDiagnosticsAnalyzerTests.cs:74]: Test for constructor parameter added.

File: src/FunFair.CodeAnalysis.Tests/ProhibitedClassesDiagnosticsAnalyzerTests.cs (hunk 2 above)
Comment (vladimirnesic1980) [reply src/FunFair.CodeAnalysis.Tests/ProhibitedClassesDiagnosticsAnalyzerTests.cs:18]: Done

**Review by credfeto — APPROVED** [review]

**Pull request merged**